plugins/outputs/whatap/DataOutputX.go

```diff
@@ -0,0 +1,225 @@
+	if b == nil || len(b) == 0 {
+		out.WriteInt(0)
+	} else {
+		out.WriteInt(int32(len(b)))
+		out.WriteBytes(b)
+	}
+	return out
+}
+
+func (out *DataOutputX) WriteBlob(value []byte) *DataOutputX {
+	if value == nil || len(value) == 0 {
+		out.WriteByte1(0)
+	} else {
+		sz := len(value)
+		if sz <= 253 {
+			out.WriteByte1(byte(sz))
+			out.WriteBytes(value)
+		} else if sz <= 65535 {
+			buff := []byte{255, 0, 0}
+			out.WriteBytes(SetBytesShort(buff, 1, int16(sz)))
+			out.WriteBytes(value)
+		} else {
+			buff := []byte{254, 0, 0, 0, 0}
+			out.WriteBytes(SetBytesInt(buff, 1, int32(sz)))
+			out.WriteBytes(value)
+		}
+	}
+	return out
+}
+func (out *DataOutputX) WriteDecimal(v int64) *DataOutputX {
+
+	switch {
+	case v == 0:
+		out.WriteByte1(0)
+	case math.MinInt8 <= v && v <= math.MaxInt8:
+		b := []byte{0, 0}
+		b[0] = 1
+		b[1] = byte(v)
+		out.WriteBytes(b)
+	case math.MinInt16 <= v && v <= math.MaxInt16:
+		b := []byte{0, 0, 0}
+		b[0] = 2
+		SetBytesShort(b, 1, int16(v))
+		out.WriteBytes(b)
+	case INT3_MIN_VALUE <= v && v <= INT3_MAX_VALUE:
+		b := []byte{0, 0, 0, 0}
+		b[0] = 3
+		out.WriteBytes(SetBytesInt3(b, 1, int32(v)))
+	case math.MinInt32 <= v && v <= math.MaxInt32:
+		b := []byte{0, 0, 0, 0, 0}
+		b[0] = 4
+		out.WriteBytes(SetBytesInt(b, 1, int32(v)))
+	case LONG5_MIN_VALUE <= v && v <= LONG5_MAX_VALUE:
+		b := []byte{0, 0, 0, 0, 0, 0}
+		b[0] = 5
+		out.WriteBytes(SetBytesLong5(b, 1, v))
+	case math.MinInt64 <= v && v <= math.MaxInt64:
+		b := []byte{0, 0, 0, 0, 0, 0, 0, 0, 0}
+		b[0] = 8
+		out.WriteBytes(SetBytesLong(b, 1, v))
+	}
+	return out
+}
+func (out *DataOutputX) WriteText(s string) *DataOutputX {
+	if s == "" {
+		out.WriteByte1(0)
+	} else {
+		out.WriteBlob([]byte(s))
+	}
+	return out
+}
+
+func (out *DataOutputX) WriteBytes(b []byte) *DataOutputX {
+	out.written += len(b)
+	out.buffer.Write(b)
+	return out
+}
+func (out *DataOutputX) WriteBool(b bool) *DataOutputX {
+	out.WriteBytes(ToBytesBool(b))
+	return out
+}
+
+// go vet : method WriteByte(b byte) (*DataOutputX)
+// should have signature WriteByte(byte) error
+func (out *DataOutputX) WriteByte1(b byte) (*DataOutputX, error) {
+	out.written++
+	out.buffer.WriteByte(b)
+	return out, nil
+}
+func (out *DataOutputX) WriteShort(b int16) *DataOutputX {
+	out.WriteBytes(ToBytesShort(b))
+	return out
+}
+func (out *DataOutputX) WriteInt(b int32) *DataOutputX {
+	out.WriteBytes(ToBytesInt(b))
+	return out
+}
+func (out *DataOutputX) WriteLong(b int64) *DataOutputX {
+	out.WriteBytes(ToBytesLong(b))
+	return out
+}
+func (out *DataOutputX) WriteFloat(b float32) *DataOutputX {
+	out.WriteBytes(ToBytesFloat(b))
+	return out
+}
+func (out *DataOutputX) WriteDouble(b float64) *DataOutputX {
+	out.WriteBytes(ToBytesDouble(b))
+	return out
+}
+func (out *DataOutputX) Size() int {
+	return out.written
+}
+
+func ToBytesBool(b bool) []byte {
+	if b {
+		return []byte{1}
+	} else {
+		return []byte{0}
+	}
+}
+
+func ToBytesShort(v int16) []byte {
```

Comment:
There are function for serializing these types of values in https://golang.org/pkg/encoding/binary/.  For larger operations it may make sense to use https://golang.org/pkg/encoding/gob/, which essentially allows you to write a struct to the socket.  See if you can make use of either of these packages.

Reply:
It is a library to support multi languages ​​(java, python, etc.). It is different from the standard serialize library specification and cannot be applied

Reply:
So this style is used so that it is more similar to the implementation you are using for other languages?  I do think we must switch to using the `binary` package here as much as possible, documenting when we cannot.  The `binary` package is well tested and uses more efficient options to serialize values.

Reply:
Modified to use binary package.
```
func ToBytesShort(v int16) []byte {
	buf := make([]byte, 2)
	binary.BigEndian.PutUint16(buf[0:], uint16(v))
	return buf
}
```